### PR TITLE
Fix export namespace of an excluded asset

### LINKED ISSUE #4216: Scope hoisting: `import *` with re-exporting does not include all exports
# 🐛 bug report

In the following case with scope hoisting enabled, using `import *` with the `@sentry/browser` package does not include all of its exports. This package re-exports quite a bit.

## 🎛 Configuration (.babelrc, package.json, cli command)

Defaults, `parcel2 build --no-minify`

## 🤔 Expected Behavior

All of the exports should be included when importing using `import *`

## 😯 Current Behavior

Only subset of exports is included.

## 💻 Code Sample

`yarn add @sentry/browser`

```js
import * as Sentry from '@sentry/browser';

Sentry.configureScope(scope => console.log('scope', scope));
console.log('Sentry', Sentry);
```

and build with `parcel2 build --no-minify index.js`

The above logs:
```
Sentry {
  Transports: {
    BaseTransport: [Function: BaseTransport],
    FetchTransport: [Function: FetchTransport],
    XHRTransport: [Function: XHRTransport]
  },
  Integrations: {
    FunctionToString: [Function: FunctionToString] { id: 'FunctionToString' },
    InboundFilters: [Function: InboundFilters] { id: 'InboundFilters' },
    GlobalHandlers: [Function: GlobalHandlers] { id: 'GlobalHandlers' },
    TryCatch: [Function: TryCatch] { id: 'TryCatch' },
    Breadcrumbs: [Function: Breadcrumbs] { id: 'Breadcrumbs' },
    LinkedErrors: [Function: LinkedErrors] { id: 'LinkedErrors' },
    UserAgent: [Function: UserAgent] { id: 'UserAgent' }
  }
}
```

...but should include `@sentry/browser`'s other exports, such as `configureScope` and `captureException`.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | v2 branch (04f2e662b571a9caba58363aa0ce3303a282e281)
| Node             | v12.16.1
| npm/Yarn         | yarn 1.21.1
| Operating System | macOS 10.15.3

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

Closes #4216

An asset with `export * from ...` might be excluded by `concat`, so the exportsSpecifier will be undefined (so `$parcel$exportWildcard(something, ReexportingExportsSpecifier)` doesn't work).

Solution: don't emit `$parcel$exportWildcard` if the imported asset is a ES6 module, in which case `bundleGraph.getExportedSymbols(mod)` are traversed and assignments emitted.

cc @wbinnssmith, does this fix your problem?

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.46s | -870.00ms 🚀 |
| Cached | 3.24s | -23.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/parcel.d5807e82.webp | 102.94kb | +0.00b | 74.00ms | -5.00ms 🚀 |
| dist/index.js | 1.07kb | +5.00b ⚠️ | 171.00ms | -6.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/parcel.d5807e82.webp | 102.94kb | +0.00b | 2.00ms | -51.00ms 🚀 |
| dist/index.js | 1.07kb | +5.00b ⚠️ | 5.00ms | -52.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.97s | -311.00ms |
| Cached | 3.40s | +13.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 67.47kb | +429.00b ⚠️ | 1.12s | +53.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 67.47kb | +429.00b ⚠️ | 9.00ms | +3.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/ak-editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.54s | -10.00ms |
| Cached | 3.26s | -103.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 743.00b | +7.00b ⚠️ | 163.00ms | +7.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 743.00b | +7.00b ⚠️ | 5.00ms | -0.00ms |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/258366141846520321)**

**Comment by devongovett** [comment]
> An asset with export * from ... might be excluded by concat

Should it have been excluded though? Seems like if later code is referencing it, then maybe excluding it was wrong? Probably misunderstanding.

**Comment by mischnic** [comment]
> Should it have been excluded though?

Well, symbols in `concat` are marked used with `bundleGraph.resolveSymbol()`, so in the example below, "exports.js" is not actually "used". 
But when calling the wildcard function, that asset's exportidentifier is needed.

```js
// index.js
export * from "./exports";

export const Main = "main";

//  exports.js
export { a } from "./version";
export { b } from "./version2";

// version(2) just export a variable
```

<br>

Another difference:
output when simply including the missing asset:
```js
(function(){
function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function (key) {
    if (key === 'default' || key === '__esModule') {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });
  return dest;
}

// ASSET: /Users/niklas/Documents/_dev/_git/parcel/parcel/packages/core/integration-tests/test/integration/scope-hoisting/es6/import-namespace-sideEffects/other/index.js
var $e367b6edc7f6de5eef03bc6e42a99$exports = {};
// ASSET: /Users/niklas/Documents/_dev/_git/parcel/parcel/packages/core/integration-tests/test/integration/scope-hoisting/es6/import-namespace-sideEffects/other/exports.js
var $d52a61ca40f5bad1d825ae05f8$exports = {};
var $e7fabb81ca3e1f1a1428a4a075efc$export$a = "foo";
var $c295ce722e44b4d388a0f05140b040$export$b = "bar";
$d52a61ca40f5bad1d825ae05f8$exports.a = $e7fabb81ca3e1f1a1428a4a075efc$export$a;
$d52a61ca40f5bad1d825ae05f8$exports.b = $c295ce722e44b4d388a0f05140b040$export$b;
$parcel$exportWildcard($e367b6edc7f6de5eef03bc6e42a99$exports, $d52a61ca40f5bad1d825ae05f8$exports);
const $e367b6edc7f6de5eef03bc6e42a99$export$Main = "main";
$e367b6edc7f6de5eef03bc6e42a99$exports.Main = $e367b6edc7f6de5eef03bc6e42a99$export$Main;
output = $e367b6edc7f6de5eef03bc6e42a99$exports;
})();
```
output with this change:
```js
(function(){
// ASSET: /Users/niklas/Documents/_dev/_git/parcel/parcel/packages/core/integration-tests/test/integration/scope-hoisting/es6/import-namespace-sideEffects/other/exports.js
// ASSET: /Users/niklas/Documents/_dev/_git/parcel/parcel/packages/core/integration-tests/test/integration/scope-hoisting/es6/import-namespace-sideEffects/other/version.js
var $e7fabb81ca3e1f1a1428a4a075efc$export$a = "foo";
var $c295ce722e44b4d388a0f05140b040$export$b = "bar";
// ASSET: /Users/niklas/Documents/_dev/_git/parcel/parcel/packages/core/integration-tests/test/integration/scope-hoisting/es6/import-namespace-sideEffects/other/index.js
var $e367b6edc7f6de5eef03bc6e42a99$exports = {};
$e367b6edc7f6de5eef03bc6e42a99$exports.a = $e7fabb81ca3e1f1a1428a4a075efc$export$a;
$e367b6edc7f6de5eef03bc6e42a99$exports.b = $c295ce722e44b4d388a0f05140b040$export$b;
const $e367b6edc7f6de5eef03bc6e42a99$export$Main = "main";
$e367b6edc7f6de5eef03bc6e42a99$exports.Main = $e367b6edc7f6de5eef03bc6e42a99$export$Main;
output = $e367b6edc7f6de5eef03bc6e42a99$exports;
})();
```

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "mocha": "^6.1.3",
     "mocha-junit-reporter": "^1.21.0",
     "mocha-multi-reporters": "^1.1.7",
+    "nyc": "^15.0.0",
```
Comment (wbinnssmith) [inline package.json]: where is this used?

Comment (mischnic) [reply package.json]: Nowhere, I tested coverage once.

File: packages/shared/scope-hoisting/src/link.js [hunk 2]
```diff
@@ -438,6 +459,86 @@ export function link({
         );
         let mod = nullthrows(bundleGraph.getDependencyResolution(dep, bundle));
         path.replaceWith(t.valueToNode(mod.id));
+      } else if (callee.name === '$parcel$exportAll') {
```
Comment (devongovett) [inline packages/shared/scope-hoisting/src/link.js]: Not sure I totally understand why this is necessary vs the old `$parcel$exportWildcard` with a `$parcelRequire` inside. Seems like a bunch of logic got duplicated?

**Comment by mischnic** [comment]
This is indeed a giant hack and way too complex, let's do https://github.com/parcel-bundler/parcel/pull/4387 instead.

**Pull request closed without merging**